torch/distributed/device_mesh.py

```diff
@@ -769,26 +593,11 @@ def __getitem__(
                 (mesh_dim_names,) if isinstance(mesh_dim_names, str) else mesh_dim_names
             )
 
-            if mesh_dim_names == self.mesh_dim_names:
-                return self
-            else:
-                slice_mesh_dims = _mesh_resources._get_slice_mesh_dims(
-                    self, mesh_dim_names
-                )
-                # When using FakeTensorMode to trace the model, `create_sub_mesh()` will
-                # fail as it will require a real tensor to manipulate.
-                # `unset_fake_temporarily()` will allow us to materialize the tensors
-                # within `_mesh_resources`, which should not affect modling.
-                #
-                # Note that this should be orthogonal to torch.compile(). But whether
-                # we can compile device_mesh `slicing` (no graph break) is not verified
-                # yet and need a follow-up,
-                # TODO: compiler + device_mesh slicing.
-                with torch._subclasses.fake_tensor.unset_fake_temporarily():
-                    submesh = _mesh_resources.create_sub_mesh(
-                        self, mesh_dim_names, slice_mesh_dims
-                    )
-                return submesh
+            return DeviceMesh.from_storage(
+                self.storage,
+                [self.storage.names_to_layouts[n] for n in mesh_dim_names],
```

Comment:
I think of a case when users have ["tp", "dp", "pp"] and a flatten between tp and dp is performed. Then slicing ["tp", "tp_dp"] is actual legit according to this, but I guess it should be not valid?

Reply:
I think you're right. More generally, I forgot to add any checks/asserts that for any given DeviceMesh its Layouts are "non-overlapping" (I'm reusing the [same terminology as Tensors](https://github.com/pytorch/pytorch/blob/db32b60662b2f2bdcad980127d5dc4b66b02a7e4/c10/core/Contiguity.h#L155-L189)).